src/main/java/org/spongepowered/api/util/AABB.java

```diff
@@ -0,0 +1,444 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.util;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.flowpowered.math.vector.Vector3d;
+import com.flowpowered.math.vector.Vector3i;
+import org.spongepowered.api.entity.EntityTypes;
+import org.spongepowered.api.entity.living.player.Player;
+import org.spongepowered.api.util.blockray.BlockRay;
+import org.spongepowered.api.util.blockray.BlockRayHit;
+import org.spongepowered.api.world.World;
+import org.spongepowered.api.world.extent.Extent;
+
+import java.util.Optional;
+
+import javax.annotation.Nullable;
+
+/**
+ * An axis aligned bounding box. That is, an un-rotated cuboid.
+ * It is represented by its minimum and maximum corners.
+ *
+ * <p>The box will never be degenerate: the corners are always not equal and
+ * respect the minimum and maximum properties.</p>
+ *
+ * <p>This class is immutable, all objects returned are either new instances or
+ * itself.</p>
+ */
+public class AABB {
+
+    private final Vector3d min;
+    private final Vector3d max;
+    @Nullable private Vector3d size = null;
+    @Nullable private Vector3d center = null;
+
+    /**
+     * Constructs a new bounding box from two opposite corners.
+     * Fails the resulting box would be degenerate (a dimension is 0).
+     *
+     * @param firstCorner The first corner
+     * @param secondCorner The second corner
+     */
+    public AABB(Vector3i firstCorner, Vector3i secondCorner) {
+        this(firstCorner.toDouble(), secondCorner.toDouble());
```

Comment:
`this(checkNotNull(firstCorner, "firstCorner").toDouble(), checkNotNull(secondCorner, "secondCorner").toDouble());`